File: src/myrsa/gen_key.py
from typing import Tuple
from myrsa.gcd import ext_euclid
from myrsa.prime_nums import produce_pq
from myrsa.types import *


def gen_keys(n_bit_size: int, n_threads: int = 4) -> Tuple[PublicKey, PrivateKey]:
    p, q = produce_pq(n_bit_size, n_threads)
    n = p * q
    phi_n = (p - 1) * (q - 1)
    e = 65537
    d, _, _ = ext_euclid(e, phi_n)
    if d < 0:
        d += phi_n

    return PublicKey(n_bit_size//2, n, e), PrivateKey(n_bit_size//2, n, d, p, q)

File: backend/rsa/gen_key.go
package rsa

import "math/big"

func GenRSAKeys(nBitSize, nGoRoutines int) (*PublicKey, *PrivateKey) {
	p, q := producePQ(nBitSize, nGoRoutines)
	n := new(big.Int).Mul(p, q)
	phiN := new(big.Int).Mul(new(big.Int).Sub(p, one), new(big.Int).Sub(q, one))
	d, _, _ := extEuclid(new(big.Int).Set(e), new(big.Int).Set(phiN))
	if d.Cmp(zero) < 0 {
		d.Add(d, phiN)
	}
	pubKey := &PublicKey{
		NLength: nBitSize,
		NValue: n,
		EValue: big.NewInt(65537),
	}
	priKey := &PrivateKey{
		NLength: nBitSize,
		NValue: n,
		DValue: d,
		PValue: p,
		QValue: q,
	}
	return pubKey, priKey
}